### PR TITLE
defuse with pointers to the statement in a variable is used/defined

### LINKED ISSUE #513: defuse chain visitor returns also a pointer to the node or the statement
with the defuse visitor we currently return define-use chains where we state how many times a certain variable is used. It could be useful to add also a pointer that indicates in which node/statement a variable is used

[ud-chain wiki](https://en.wikipedia.org/wiki/Use-define_chain#:~:text=From%20Wikipedia%2C%20the%20free%20encyclopedia,some%20value%20to%20a%20variable.)

### PR DESCRIPTION
**Pull request by cattabiani**
In this PR we expand DUInstance to also return the ... "line" in which defuse is used/defined. In particular, we return a pointer to the outermost `binary_expression`. Why not returning an `expression_statement` pointer? In this way we handle these 2 cases:

- `IF (tau == 0)` \\ not expression statement, binary expression
- `x = x + 1`  \\ binary expression and expression statement

This is propedeutic for using the defuse visitor inside #495 .

Addresses #513 

PS: I also included a few corner cases in the unit tests for defuse that were not checked

### CODE REVIEW

File: test/unit/visitor/defuse_analyze.cpp [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 #include <catch/catch.hpp>
 
-#include "ast/program.hpp"
+#include "ast/all.hpp"
```
Comment (pramodk) [inline test/unit/visitor/defuse_analyze.cpp:10]: nitpik : if we have "few" ast classes to use then we include individual headers and avoid `all.hpp` (which is very large).

**Review by pramodk — APPROVED** [review]
I was initially wondering about the same:

>  In particular, we return a pointer to the outermost binary_expression. Why not returning an expression_statement pointer?

but by looking at test examples and use cases, it seems make sense.

Overall, looks nicely done with minimal changes / extensions! 👌
